src/App/Scan.hs

```diff
@@ -2,71 +2,77 @@
-import Effect.Logger
-import Effect.ReadFS (ReadFSErr(..))
-import qualified Strategy.Carthage as Carthage
-import qualified Strategy.Cocoapods.Podfile as Podfile
-import qualified Strategy.Cocoapods.PodfileLock as PodfileLock
-import qualified Strategy.Go.GoList as GoList
-import qualified Strategy.Go.Gomod as Gomod
-import qualified Strategy.Go.GopkgLock as GopkgLock
-import qualified Strategy.Go.GopkgToml as GopkgToml
-import qualified Strategy.Go.GlideLock as GlideLock
-import qualified Strategy.Gradle as Gradle
-import qualified Strategy.Maven.Pom as MavenPom
-import qualified Strategy.Maven.PluginStrategy as MavenPlugin
-import qualified Strategy.Node.NpmList as NpmList
-import qualified Strategy.Node.NpmLock as NpmLock
-import qualified Strategy.Node.PackageJson as PackageJson
-import qualified Strategy.Node.YarnLock as YarnLock
-import qualified Strategy.NuGet.PackagesConfig as PackagesConfig
-import qualified Strategy.NuGet.PackageReference as PackageReference
-import qualified Strategy.NuGet.ProjectAssetsJson as ProjectAssetsJson
-import qualified Strategy.NuGet.ProjectJson as ProjectJson
-import qualified Strategy.NuGet.Nuspec as Nuspec
-import qualified Strategy.Python.Pipenv as Pipenv
-import qualified Strategy.Python.ReqTxt as ReqTxt
-import qualified Strategy.Python.SetupPy as SetupPy
-import qualified Strategy.Ruby.BundleShow as BundleShow
-import qualified Strategy.Ruby.GemfileLock as GemfileLock
-import Types
+import Network.HTTP.Req (HttpException)
+
+import VPSScan.RunSherlock
+import VPSScan.ScotlandYard
+import VPSScan.RunIPR
 
 data ScanCmdOpts = ScanCmdOpts
   { cmdBasedir :: FilePath
   , cmdDebug   :: Bool
   , cmdOutFile :: Maybe FilePath
+  , scanVpsOpts :: VPSOpts
+  } deriving (Eq, Ord, Show, Generic)
+
+data VPSOpts = VPSOpts
+  { vpsSherlock :: SherlockOpts
+  , vpsIpr :: IPROpts
+  , vpsScotlandYard :: ScotlandYardOpts
   } deriving (Eq, Ord, Show, Generic)
 
 scanMain :: ScanCmdOpts -> IO ()
-scanMain ScanCmdOpts{..} = do
-  hSetBuffering stdout NoBuffering
-  hSetBuffering stderr NoBuffering
+scanMain opts@ScanCmdOpts{..} = do
   basedir <- validateDir cmdBasedir
-
-  scan basedir cmdOutFile
-    & withLogger (bool SevInfo SevDebug cmdDebug)
-    & runThreaded
+  result <- runError @VPSError $ runScotlandYard $ runSherlock $ runIPR $ vpsScan basedir opts
+  case result of
+    Left err -> do
+      print err
+      exitFailure
+    Right _ -> pure ()
+
+----- main logic
+
+data VPSError
+  = IPRFailed IPRError
+  | SherlockFailed SherlockError
+  | Couldn'tGetScanId HttpException
+  | Couldn'tUpload HttpException
+  deriving (Show, Generic)
+
+vpsScan ::
+  ( Has ScotlandYard sig m
+  , Has IPR sig m
+  , Has Sherlock sig m
+  , Has (Error VPSError) sig m
+  ) => Path Abs Dir -> ScanCmdOpts -> m ()
+vpsScan basedir ScanCmdOpts{..} = do
+  let VPSOpts{..} = scanVpsOpts
+  response <- tagError Couldn'tGetScanId =<< createScotlandYardScan vpsScotlandYard
+ 
+  traceM $ "Running scan on directory " ++ show basedir
```

Comment:
rather than using `traceM` here, let's use the `Trace` effect, which gives us:
```hs
trace :: Has Trace sig m => String -> m ()
```

it has a few carriers/interpreters, exported from these three modules:
- `Control.Carrier.Trace.Printing` - this mirrors the behavior of `traceM`
- `Control.Carrier.Trace.Ignoring` - this ignores the traces
- `Control.Carrier.Trace.Returning` - this collects the trace messages as a list and returns them out of the computation (most often useful for testing)

you'll have to use `runTrace` (imported from one of the carrier modules) at the callsite of vpsScan

Reply:
I knew using `traceM` in production code was a bad idea. Meant to ask you about that. Thanks!